lib/components/map/enhanced-stop-marker.js

```diff
@@ -0,0 +1,116 @@
+import { divIcon } from 'leaflet'
+import React, { Component } from 'react'
+import ReactDOMServer from 'react-dom/server'
+import { connect } from 'react-redux'
+import { Marker, Popup } from 'react-leaflet'
+import styled from 'styled-components'
+
+import * as mapActions from '../../actions/map'
+import * as uiActions from '../../actions/ui'
+import { getStopName } from '../../util/viewer'
+
+const { getTextWidth } = coreUtils.itinerary
+
+const BaseStopIcon = styled.div`
+  background: #fff;
+  border: ${props => props.active ? '#000 3px' : '#333 1px'} solid;
+  border-radius: 17px;
+  color: #000;
+  font-size: 16px;
+  font-weight: bold;
+  height: 12px;
+  line-height: 0px;
+  padding-left: 7px;
+  padding-top: 12px;
+  width: ${props => `${props.width || 12}px`}
+`
+
+class EnhancedStopMarker extends Component {
+  onClickView = () => {
+    const { setViewedStop, stop } = this.props
+    setViewedStop({ stopId: stop.id })
+  }
+
+  onFromClick = () => {
+    this.setLocation('from')
+  }
+
+  onToClick = () => {
+    this.setLocation('to')
+  }
+
+  setLocation (locationType) {
+    const { setLocation, stop } = this.props
+    const { lat, lon, name } = stop
+    setLocation({ location: { lat, lon, name }, locationType })
+  }
+
+  render () {
+    const { activeStopId, languageConfig, stop } = this.props
+    const { code, id, lat, lon, name } = stop
+    const [, stopId] = id.split(':')
+    const stopCode = code || stopId
+    const routeShortNames = stop.routes?.map(r => r.shortName).join(' ')
+    const stopLabel = routeShortNames || ''
+    if (!stopLabel) return null
+    const width = getTextWidth(stopLabel) * 0.83 + 80 / getTextWidth(stopLabel)
+    const icon = divIcon({
+      className: '',
+      html: ReactDOMServer.renderToStaticMarkup(
+        <BaseStopIcon
+          active={id === activeStopId}
+          // Show actual stop name on hover for easy identification.
+          title={getStopName(stop)}
+          width={width}
+        >
+          {stopLabel}
+        </BaseStopIcon>
+      ),
+      iconAnchor: [(width / 2), 10]
+    })
+    return (
+      <Marker icon={icon} position={[lat, lon]}>
+        <Popup>
+          <BaseMapStyled.MapOverlayPopup>
+            <BaseMapStyled.PopupTitle>{name}</BaseMapStyled.PopupTitle>
+            <BaseMapStyled.PopupRow>
+              <span>
+                <b>Stop ID:</b> {stopCode}
+              </span>
+              <button onClick={this.onClickView}>
+                {languageConfig.stopViewer || 'Stop Viewer'}
+              </button>
+            </BaseMapStyled.PopupRow>
+
+            {/* The 'Set as [from/to]' ButtonGroup */}
+            <BaseMapStyled.PopupRow>
+              <b>Plan a trip:</b>
+              <FromToLocationPicker
+                onFromClick={this.onFromClick}
+                onToClick={this.onToClick}
+              />
+            </BaseMapStyled.PopupRow>
+          </BaseMapStyled.MapOverlayPopup>
+        </Popup>
```

Comment:
This is all basically duplicated code of https://github.com/opentripplanner/otp-ui/blob/b6da184309a23bd31903b110f7b54201761a8b76/packages/stops-overlay/src/default-stop-marker.js#L48-L69. Perhaps there is some kind of opportunity to share this common code.

Reply:
Grant mentioned that they're importing a bunch of styled.js stuff from `@opentripplanner-*/lib/styled`, so I think we should go ahead and export all of the styled exports directly from `index.js` within various otp-ui packages. I'll work on a PR there.